[libc++] Fix the locale base API on Linux with musl

androm3da opened the pull request:
Since `363bfd6090b0 ([libc++] Use the new locale base API on Linux (#128007), 2025-02-24)`, musl targets will fail to build with errors like the below:

    In file included from /[b/f/w/src/git/out/stage2/include/c++/v1/__locale:14](https://cs.corp.google.com/piper///depot/google3/b/f/w/src/git/out/stage2/include/c%2B%2B/v1/__locale?l=14):
    In file included from /[b/f/w/src/git/out/stage2/include/c++/v1/__locale_dir/locale_base_api.h:123](https://cs.corp.google.com/piper///depot/google3/b/f/w/src/git/out/stage2/include/c%2B%2B/v1/__locale_dir/locale_base_api.h?l=123):
    /[b/f/w/src/git/out/stage2/include/c++/v1/__locale_dir/support/linux.h:98](https://cs.corp.google.com/piper///depot/google3/b/f/w/src/git/out/stage2/include/c%2B%2B/v1/__locale_dir/support/linux.h?l=98):12: error: no member named 'strtoll_l' in the global namespace
       98 |   return ::strtoll_l(__nptr, __endptr, __base, __loc);
          |          ~~^
    /[b/f/w/src/git/out/stage2/include/c++/v1/__locale_dir/support/linux.h:103](https://cs.corp.google.com/piper///depot/google3/b/f/w/src/git/out/stage2/include/c%2B%2B/v1/__locale_dir/support/linux.h?l=103):10: error: no member named 'strtoull_l' in the global namespace; did you mean '__strtoull'?
      103 |   return ::strtoull_l(__nptr, __endptr, __base, __loc);
          |          ^~~~~~~~~~~~
          |          __strtoull
    /[b/f/w/src/git/out/stage2/include/c++/v1/__locale_dir/support/linux.h:102](https://cs.corp.google.com/piper///depot/google3/b/f/w/src/git/out/stage2/include/c%2B%2B/v1/__locale_dir/support/linux.h?l=102):1: note: '__strtoull' declared here
      102 | __strtoull(const char* __nptr, char** __endptr, int __base, __locale_t __loc) {
          | ^
    2 errors generated.

pirama-arumuga-nainar committed:
[libcxx] avoid the "_l" calls on musl platforms

Since 363bfd6090b0 ([libc++] Use the new locale base API on Linux (#128007), 2025-02-24),
musl targets will fail to build with errors like the below:

    In file included from /[b/f/w/src/git/out/stage2/include/c++/v1/__locale:14](https://cs.corp.google.com/piper///depot/google3/b/f/w/src/git/out/stage2/include/c%2B%2B/v1/__locale?l=14):
    In file included from /[b/f/w/src/git/out/stage2/include/c++/v1/__locale_dir/locale_base_api.h:123](https://cs.corp.google.com/piper///depot/google3/b/f/w/src/git/out/stage2/include/c%2B%2B/v1/__locale_dir/locale_base_api.h?l=123):
    /[b/f/w/src/git/out/stage2/include/c++/v1/__locale_dir/support/linux.h:98](https://cs.corp.google.com/piper///depot/google3/b/f/w/src/git/out/stage2/include/c%2B%2B/v1/__locale_dir/support/linux.h?l=98):12: error: no member named 'strtoll_l' in the global namespace
       98 |   return ::strtoll_l(__nptr, __endptr, __base, __loc);
          |          ~~^
    /[b/f/w/src/git/out/stage2/include/c++/v1/__locale_dir/support/linux.h:103](https://cs.corp.google.com/piper///depot/google3/b/f/w/src/git/out/stage2/include/c%2B%2B/v1/__locale_dir/support/linux.h?l=103):10: error: no member named 'strtoull_l' in the global namespace; did you mean '__strtoull'?
      103 |   return ::strtoull_l(__nptr, __endptr, __base, __loc);
          |          ^~~~~~~~~~~~
          |          __strtoull
    /[b/f/w/src/git/out/stage2/include/c++/v1/__locale_dir/support/linux.h:102](https://cs.corp.google.com/piper///depot/google3/b/f/w/src/git/out/stage2/include/c%2B%2B/v1/__locale_dir/support/linux.h?l=102):1: note: '__strtoull' declared here
      102 | __strtoull(const char* __nptr, char** __endptr, int __base, __locale_t __loc) {
          | ^
    2 errors generated.

ldionne commented on libcxx/include/__locale_dir/support/linux.h:
```diff
@@ -95,12 +95,20 @@ inline _LIBCPP_HIDE_FROM_ABI long double __strtold(const char* __nptr, char** __
 inline _LIBCPP_HIDE_FROM_ABI long long __strtoll(const char* __nptr, char** __endptr, int __base, __locale_t __loc) {
+#if !defined(_LIBCPP_HAS_MUSL_LIBC)
   return ::strtoll_l(__nptr, __endptr, __base, __loc);
+#else
+  return ::strtoll(__nptr, __endptr, __base);
+#endif
```
```suggestion
#else
  (void)__loc;
  return ::strtoll(__nptr, __endptr, __base);
#endif
```

ldionne commented on libcxx/include/__locale_dir/support/linux.h:
```diff
@@ -95,12 +95,20 @@ inline _LIBCPP_HIDE_FROM_ABI long double __strtold(const char* __nptr, char** __
 __strtoull(const char* __nptr, char** __endptr, int __base, __locale_t __loc) {
+#if !defined(_LIBCPP_HAS_MUSL_LIBC)
   return ::strtoull_l(__nptr, __endptr, __base, __loc);
+#else
+  return ::strtoull(__nptr, __endptr, __base);
+#endif
```
```suggestion
#else
  (void)__loc;
  return ::strtoull(__nptr, __endptr, __base);
#endif
```

ldionne reviewed (approved):
LGTM, thanks for the fix and sorry for the slow reaction time.

Do you know if this means we can actually get rid of this code path in `locale_base_api.h`?

```
#  elif defined(__wasi__) || _LIBCPP_HAS_MUSL_LIBC
#    include <__locale_dir/locale_base_api/musl.h>
```

ldionne committed:
Apply suggestions from code review

ldionne commented:
This can be landed as soon as the CI is green. If someone lands this before I do, please edit the commit message to avoid butchering `git log`.

Thanks for the fix @androm3da

androm3da committed:
Use "_LIBCPP_HAS_MUSL_LIBC" directly

Instead of `#if defined(_LIBCPP_HAS_MUSL_LIBC)`, we can rely on
`#if _LIBCPP_HAS_MUSL_LIBC` directly.

This should address these failures:

	  # .---command stdout------------
	  # | /__w/llvm-project/llvm-project/build/frozen-cxx03-headers/libcxx/test-suite-install/include/c++/v1/__locale_dir/support/linux.h:98:6: error: '_LIBCPP_HAS_MUSL_LIBC' is always defined to 1 or 0. [libcpp-internal-ftms,-warnings-as-errors]
	  # |    98 | #if !defined(_LIBCPP_HAS_MUSL_LIBC)
	  # |       |      ^
	  # | /__w/llvm-project/llvm-project/build/frozen-cxx03-headers/libcxx/test-suite-install/include/c++/v1/__locale_dir/support/linux.h:108:6: error: '_LIBCPP_HAS_MUSL_LIBC' is always defined to 1 or 0. [libcpp-internal-ftms,-warnings-as-errors]
	  # |   108 | #if !defined(_LIBCPP_HAS_MUSL_LIBC)
	  # |       |      ^
	  # `-----------------------------

androm3da commented:
> This can be landed as soon as the CI is green. If someone lands this before I do, please edit the commit message to avoid butchering `git log`.
> 
> Thanks for the fix @androm3da

Note that this fix was actually proposed by @pirama-arumuga-nainar. 

Even though @pirama-arumuga-nainar is the author of the first commit in this PR I think GH will take my name as the author when it's merged.  So I'll add a `Co-authored-by` to the message if that's satisfactory @pirama-arumuga-nainar

androm3da commented:
> please edit the commit message to avoid butchering git log.

Will GH catenate the commit messages in the squash-and-merge?  In which case I'm happy to edit it so that it reflects the original commit message in the first commit.  Is that what you meant, or what butchering specifically should I try to avoid?

androm3da commented:
> LGTM, thanks for the fix and sorry for the slow reaction time.
> 
> Do you know if this means we can actually get rid of this code path in `locale_base_api.h`?
> 
> ```
> #  elif defined(__wasi__) || _LIBCPP_HAS_MUSL_LIBC
> #    include <__locale_dir/locale_base_api/musl.h>
> ```

This change shouldn't impact `__wasi__` targets.  Should we include `__wasi__` in these changes to `__strtoll()`, `__strtoull()`?

pirama-arumuga-nainar commented:
> > LGTM, thanks for the fix and sorry for the slow reaction time.
> > Do you know if this means we can actually get rid of this code path in `locale_base_api.h`?
> > ```
> > #  elif defined(__wasi__) || _LIBCPP_HAS_MUSL_LIBC
> > #    include <__locale_dir/locale_base_api/musl.h>
> > ```
> 
> This change shouldn't impact `__wasi__` targets. Should we include `__wasi__` in these changes to `__strtoll()`, `__strtoull()`?

The separate definitions are still needed for `wasi` - it's not a Linux environment.  .  We could rename the file to `wasi.h` to make this explicit.

> So I'll add a Co-authored-by to the message if that's satisfactory @pirama-arumuga-nainar

Thanks for the credit, but no need to go out of the way for this.  I'm happy to just have the breakage fixed.

androm3da commented:
I'll rebase this after #128950 lands to see if that addresses the CI issues.

ldionne commented:
Let's not wait to merge this. The FreeBSD failure is a fluke that will be fixed by https://github.com/llvm/llvm-project/pull/128950.

As for #121795, it can remove that code when it lands, but let's get `main` out of a broken state.

philnik777 commented:
> Let's not wait to merge this. The FreeBSD failure is a fluke that will be fixed by #128950.
> 
> As for #121795, it can remove that code when it lands, but let's get `main` out of a broken state.

Just to be clear: I didn't mean that #121795 should block this. I found it just ironic that if I landed the patch a bit earlier this never would have happened.
